### PR TITLE
Feature/processing data models

### PR DESCRIPTION
**Pull request by tsohleDev**
# M in MVC

in this PR I:
-  Create model classes for all entities as shown in the [ERD diagram](../images/blog_app_erd_v1_1.png).
-  Set up associations between models.
    - Ensure that `author_id` column in `posts` table is the foreign_key for the `users` table.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: :author_id, dependent: :destroy
+  has_many :comments, foreign_key: :author_id, dependent: :destroy
+  has_many :likes, foreign_key: :author_id, dependent: :destroy
+end
```
Comment (franclobo) [inline app/models/user.rb]: - [ ] This module requires that the User model has a method that returns the 3 most recent posts. This should be similar to the example below:

```ruby
def update_post_counter
  author.increment!(:posts_counter)
end
```

File: app/models/post.rb
Comment (franclobo) [inline app/models/post.rb:1]: - [ ] This module requires that the Post model has a method that updates the posts counter for a user. This should be similar to the example below:

```ruby
def update_post_counter
  author.increment!(:posts_counter)
end
```

- [ ] Post model should have a method that returns the 5 most recent comments for a given post. This should be similar to the example below:

```ruby
def recent_comments
  comments.order('created_at Desc').limit(5)
end
```

- **[OPTIONAL]** It is a nice idea to use one of the Rails Active Callbacks after_save to call the update_posts_counter method every time a new like has been saved to the database, then make the update_posts_counter` method private, similar to:

```ruby
after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
+  belongs_to :post, counter_cache: :likes_counter
+end
```
Comment (franclobo) [inline app/models/like.rb]: - [ ] The Like model should have a method that updates the likes counter for a post. This should be similar to the example below:

```ruby
def update_likes_counter
  post.increment!(:likes_counter)
end
```

- **[OPTIONAL]** It is a nice idea to use one of the Rails Active Callbacks after_save to call the update_likes_counter method every time a new like has been saved to the database, then make the update_likes_counter` method private, similar to:

```ruby
  after_save :update_like_counter

  private

  def update_like_counter
    author.increment!(:likes_counter)
  end
```

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
+  belongs_to :post, counter_cache: :comments_counter
+end
```
Comment (franclobo) [inline app/models/comment.rb]: - [ ] Comment model should have a method that updates the comments counter for a post. This should be similar to the example below:

```ruby
def update_comments_counter
  post.increment!(:comments_counter)
end
```

- **[OPTIONAL]** It is a nice idea to use one of the Rails Active Callbacks after_save to call the update_comments_counter method every time a new like has been saved to the database, then make the update_comments_counter` method private, similar to:

```ruby
  after_save :update_comments_counter

  private

  def update_comments_counter
    author.increment!(:comments_counter)
  end
```

**Review by franclobo — CHANGES_REQUESTED** [review]
Hi @tsohleDev, 

# ♻️ Required Changes ♻️

so-close-jim-halpert

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

- No linter errors ✔️ 
- Descriptive PR ✔️ 
- Correct relationship between tables ✔️ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by tsohleDev** [comment]
Thanks @franclobo 👍

File: app/models/post.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
+  has_many :comments, dependent: :destroy, counter_cache: :comments_counter
+  has_many :likes, dependent: :destroy, counter_cache: :likes_counter
+
+  before_create :increment_user_postcounter
+  before_destroy :decrement_user_postcounter
+
```
Comment (Reem-lab) [inline app/models/post.rb:8]: - Kudos for using  one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) ✨🔥✅

**Review by Reem-lab — APPROVED** [review]
# Hi @tsohleDev,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights
- [ ] Great job on the funtionality for processing data into the models💯
- [ ] Followed Gitflow 🔥
- [ ] No Linters error ✅



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 4 above)
Comment (tsohleDev) [reply app/models/post.rb:8]: ohh Thank you @Reem-lab 😄 Kudos indeed

**Pull request merged**